### ISSUE TITLE
Potential bug on InstructionSimplify.cpp/foldSelectWithBinaryOp

### ISSUE THREAD

**Issue by glyh** [llvm/llvm-project#118815]
I am pretty new to LLVM, so what I said maybe nonsense so please forgive me if I've done such mistakes. 

https://github.com/llvm/llvm-project/blob/f1d81dbd0573502134270495aa43e3ffb3193ab0/llvm/lib/Analysis/InstructionSimplify.cpp#L100-L112

This is a simple instruction inline optimization that is **not referred anywhere in the codebase**.

But I don't see why it's the case one of X/Y need to be identical to one of TV/FV.

Isn't it the case `(TV == FV and B) ? TV : FV` always evaluating to `FV`; And `(TV != FV or B) ? TV : FV` always evaluating to `TV`, regardless of what B being?

Also, what if B contains side effects that assigns to either TV/FV?

```llvm
@x = global i32 3
@y = global i32 4

declare i32 @set_x(i32)

define void @main() {
entry:
  %A_val = icmp ne i32 @x, @y

  %set_x_val = call i32 @set_x(i32 10)  ; Call set_x, which modifies x
  %B_val = icmp ne i32 100, %set_x_val

  %C_val = or i1 %A_val, %B_val
  %D_val = select i1 %C_val, i32 @x, i32 @y
  ret void
}

define i32 @set_x(i32 %val) {
  store i32 %val, i32* @x
  ret i32 %val
}
```

**Comment by nikic** [comment]
> But I don't see why it's the case one of X/Y need to be identical to one of TV/FV.

I agree that the check for X/Y here is unnecessary: https://alive2.llvm.org/ce/z/5rJyvI

> Also, what if B contains side effects that assigns to either TV/FV?

This transform does not work on memory, there is no such thing as a side effect on SSA values.

**Comment by glyh** [comment]
So a rivival of https://github.com/llvm/llvm-project/pull/76363 and merge maybe desired?

**Comment by nikic** [comment]
We can likely just delete this transform, I think it has been subsumed by https://github.com/llvm/llvm-project/pull/76621.

**Comment by glyh** [comment]
Can I make a PR for this then?

**Comment by nikic** [comment]
Feel free :)

**Comment by glyh** [comment]
Closed #118913
